File: ios/App/OpenPPP2/Runtime/RuntimeControls.swift
import Foundation
import NetworkExtension

public enum RuntimeConnectionAction: Equatable, Sendable {
    case start
    case cancel
    case stop
    case retry
    case forceStop
    case none
}

public struct RuntimeControlState: Equatable, Sendable {
    public var action: RuntimeConnectionAction
    public var buttonEnabled: Bool
    public var buttonTitleKey: String
    public var statusTitleKey: String
    public var detailKey: String
    public var configEditable: Bool
    public var isConnected: Bool
    public var isBusy: Bool
    public var diagnosticsAvailable: Bool

    public init(
        action: RuntimeConnectionAction,
        buttonEnabled: Bool,
        buttonTitleKey: String,
        statusTitleKey: String,
        detailKey: String,
        configEditable: Bool,
        isConnected: Bool = false,
        isBusy: Bool = false,
        diagnosticsAvailable: Bool = false
    ) {
        self.action = action
        self.buttonEnabled = buttonEnabled
        self.buttonTitleKey = buttonTitleKey
        self.statusTitleKey = statusTitleKey
        self.detailKey = detailKey
        self.configEditable = configEditable
        self.isConnected = isConnected
        self.isBusy = isBusy
        self.diagnosticsAvailable = diagnosticsAvailable
    }
}

public func controlsFor(_ phase: RuntimePhase) -> RuntimeControlState {
    switch phase {
    case .idle:
        return RuntimeControlState(
            action: .start,
            buttonEnabled: true,
            buttonTitleKey: "home.connect",
            statusTitleKey: "home.notConnected",
            detailKey: "home.ready",
            configEditable: true
        )
    case .starting, .preparingHost, .connecting, .handshaking, .applyingPolicy:
        return RuntimeControlState(
            action: .cancel,
            buttonEnabled: true,
            buttonTitleKey: "common.cancel",
            statusTitleKey: "home.connecting",
            detailKey: "home.vpnStarting",
            configEditable: false,
            isBusy: true
        )
    case .connected:
        return RuntimeControlState(
            action: .stop,
            buttonEnabled: true,
            buttonTitleKey: "home.stop",
            statusTitleKey: "home.connected",
            detailKey: "",
            configEditable: false,
            isConnected: true
        )
    case .reconnecting:
        return RuntimeControlState(
            action: .stop,
            buttonEnabled: true,
            buttonTitleKey: "home.stop",
            statusTitleKey: "home.reconnecting",
            detailKey: "home.networkChanged",
            configEditable: false,
            isBusy: true
        )
    case .stopping:
        return RuntimeControlState(
            action: .none,
            buttonEnabled: false,
            buttonTitleKey: "home.stop",
            statusTitleKey: "home.disconnecting",
            detailKey: "home.stopping",
            configEditable: false,
            isBusy: true
        )
    case .failed:
        return RuntimeControlState(
            action: .retry,
            buttonEnabled: true,
            buttonTitleKey: "home.retry",
            statusTitleKey: "home.connectFailed",
            detailKey: "home.ready",
            configEditable: true
        )
    case .unknown:
        return RuntimeControlState(
            action: .forceStop,
            buttonEnabled: true,
            buttonTitleKey: "home.forceStop",
            statusTitleKey: "vpn.unknown",
            detailKey: "settings.section.diagnostics",
            configEditable: false,
            diagnosticsAvailable: true
        )
    }
}

public func runtimeSnapshot(
    previous: RuntimeSnapshot,
    status: NEVPNStatus,
    linkState: Int
) -> RuntimeSnapshot {
    RuntimeSnapshot(
        generation: previous.generation + 1,
        monotonicMs: nextMonotonicMs(after: previous.monotonicMs),
        phase: runtimePhase(status: status, linkState: linkState)
    )
}

public func runtimeSnapshot(
    previous: RuntimeSnapshot,
    phase: RuntimePhase
) -> RuntimeSnapshot {
    RuntimeSnapshot(
        generation: previous.generation + 1,
        monotonicMs: nextMonotonicMs(after: previous.monotonicMs),
        phase: phase
    )
}

private func nextMonotonicMs(after previous: UInt64) -> UInt64 {
    let now = UInt64(Date().timeIntervalSince1970 * 1000)
    return now <= previous ? previous + 1 : now
}

private func runtimePhase(status: NEVPNStatus, linkState: Int) -> RuntimePhase {
    // ponytail: this collapses NEVPNStatus + linkState until native
    // RuntimeSnapshot JSON streams to mobile; replace with direct snapshot decode.
    switch status {
    case .invalid, .disconnected:
        return .idle
    case .disconnecting:
        return .stopping
    case .reasserting:
        return .reconnecting
    case .connecting, .connected:
        return runtimePhaseForActiveLink(status: status, linkState: linkState)
    @unknown default:
        return .unknown
    }
}

private func runtimePhaseForActiveLink(status: NEVPNStatus, linkState: Int) -> RuntimePhase {
    switch linkState {
    case 0:
        return .connected
    case 2, 3, 6:
        return .preparingHost
    case 4:
        return .reconnecting
    case 5:
        return .handshaking
    default:
        return status == .connected ? .connected : .connecting
    }
}

File: android/lib/runtime/runtime_controls.dart
import '../vpn_service.dart';
import 'runtime_snapshot.dart';

enum RuntimeConnectionAction { start, cancel, stop, retry, forceStop, none }

class RuntimeControlState {
  const RuntimeControlState({
    required this.action,
    required this.buttonEnabled,
    required this.buttonLabel,
    required this.statusLabel,
    required this.detailLabel,
    required this.configEditable,
    this.isConnected = false,
    this.isBusy = false,
    this.diagnosticsAvailable = false,
  });

  final RuntimeConnectionAction action;
  final bool buttonEnabled;
  final String buttonLabel;
  final String statusLabel;
  final String detailLabel;
  final bool configEditable;
  final bool isConnected;
  final bool isBusy;
  final bool diagnosticsAvailable;
}

RuntimeControlState controlsFor(RuntimePhase phase) {
  switch (phase) {
    case RuntimePhase.idle:
      return const RuntimeControlState(
        action: RuntimeConnectionAction.start,
        buttonEnabled: true,
        buttonLabel: '连接',
        statusLabel: '未连接',
        detailLabel: '准备连接',
        configEditable: true,
      );
    case RuntimePhase.starting:
    case RuntimePhase.preparingHost:
    case RuntimePhase.connecting:
    case RuntimePhase.handshaking:
    case RuntimePhase.applyingPolicy:
      return const RuntimeControlState(
        action: RuntimeConnectionAction.cancel,
        buttonEnabled: true,
        buttonLabel: '取消',
        statusLabel: '连接中...',
        detailLabel: 'VPN 正在启动',
        configEditable: false,
        isBusy: true,
      );
    case RuntimePhase.connected:
      return const RuntimeControlState(
        action: RuntimeConnectionAction.stop,
        buttonEnabled: true,
        buttonLabel: '停止',
        statusLabel: '已连接',
        detailLabel: '',
        configEditable: false,
        isConnected: true,
      );
    case RuntimePhase.reconnecting:
      return const RuntimeControlState(
        action: RuntimeConnectionAction.stop,
        buttonEnabled: true,
        buttonLabel: '停止',
        statusLabel: '重连中...',
        detailLabel: '网络已变化',
        configEditable: false,
        isBusy: true,
      );
    case RuntimePhase.stopping:
      return const RuntimeControlState(
        action: RuntimeConnectionAction.none,
        buttonEnabled: false,
        buttonLabel: '停止',
        statusLabel: '断开中...',
        detailLabel: '正在停止 VPN',
        configEditable: false,
        isBusy: true,
      );
    case RuntimePhase.failed:
      return const RuntimeControlState(
        action: RuntimeConnectionAction.retry,
        buttonEnabled: true,
        buttonLabel: '重试',
        statusLabel: '连接失败',
        detailLabel: '检查配置后重试',
        configEditable: true,
      );
    case RuntimePhase.unknown:
      return const RuntimeControlState(
        action: RuntimeConnectionAction.forceStop,
        buttonEnabled: true,
        buttonLabel: '强制停止',
        statusLabel: '未知状态',
        detailLabel: '请查看诊断信息',
        configEditable: false,
        diagnosticsAvailable: true,
      );
  }
}

RuntimeSnapshot runtimeSnapshotFromVpnState({
  required RuntimeSnapshot previous,
  required VpnState vpnState,
  required int linkState,
}) {
  final nowMs = DateTime.now().millisecondsSinceEpoch;
  final monotonicMs =
      nowMs <= previous.monotonicMs ? previous.monotonicMs + 1 : nowMs;
  return RuntimeSnapshot(
    generation: previous.generation + 1,
    monotonicMs: monotonicMs,
    phase: _phaseFromVpn(vpnState, linkState),
  );
}

RuntimeSnapshot runtimeSnapshotForPhase({
  required RuntimeSnapshot previous,
  required RuntimePhase phase,
}) {
  final nowMs = DateTime.now().millisecondsSinceEpoch;
  return RuntimeSnapshot(
    generation: previous.generation + 1,
    monotonicMs: nowMs <= previous.monotonicMs ? previous.monotonicMs + 1 : nowMs,
    phase: phase,
  );
}

RuntimePhase _phaseFromVpn(VpnState vpnState, int linkState) {
  // ponytail: this collapses platform state + linkState until native
  // RuntimeSnapshot JSON streams to mobile; replace with direct snapshot decode.
  switch (vpnState) {
    case VpnState.disconnected:
      return RuntimePhase.idle;
    case VpnState.disconnecting:
      return RuntimePhase.stopping;
    case VpnState.connecting:
    case VpnState.connected:
      return _phaseFromActiveLink(vpnState, linkState);
  }
}

RuntimePhase _phaseFromActiveLink(VpnState vpnState, int linkState) {
  switch (linkState) {
    case 0:
      return RuntimePhase.connected;
    case 2:
    case 3:
    case 6:
      return RuntimePhase.preparingHost;
    case 4:
      return RuntimePhase.reconnecting;
    case 5:
      return RuntimePhase.handshaking;
    default:
      return vpnState == VpnState.connected
          ? RuntimePhase.connected
          : RuntimePhase.connecting;
  }
}
